### PR TITLE
EMP-2850 Introducing the Image Helper script

### PR DESCRIPTION
**Pull request by sharma-tapas**
1. Script Setup and Error Handling:
    - Uses set -eu for strict error handling
    - Adds set -o pipefail for bash
    - Sets a detailed debug prompt with source file and line number

2. Image Management:
    - Defines a list of container images from various registries (Kubernetes, Docker, Quay)
    - Supports pulling, pushing, and managing these images
 
3. Logging Functions:
    - Implements color-coded logging with debug(), info(), warn(), and fail() functions
    - Uses tput for terminal color control

4. Key Functions:
    - pull_images(): Pulls all defined container images
    - push_images_registry(): Pushes images to a specified registry
    - import_secrets(): Creates a Kubernetes secret from container registry authentication

5. CLI Interface:
    - Supports commands: pull, push, import-secrets
    - Offers dry-run mode with -n/--dry-run
    - Provides verbose output with -v
    - Includes usage help with -h/--help

6. Flexible Container Runtime Support:
    - Detects and uses appropriate authentication files

7. Zsh Compatibility Additions:
     - Added shell detection for zsh and bash
    - Used setopt for zsh-specific settings to match sh/bash behavior
    - Replaced $@ with $* for better cross-shell compatibility
    - Used "$*" and "$@" with quotes to handle arguments correctly
    - Added shellcheck disable comment for intentional word splitting

8. Spinner Animation:
    - Created a spinner() function that displays a spinning animation
    - Uses |/-\ characters for the spinner
    - Shows a progress indicator during image pulling and pushing
    - Captures command output to a temporary file
    - Provides error logging if image operations fail

9. Enhanced Error Handling:
    - Improved error capturing during image pull and push operations
    - Added more robust temporary file handling
    - Maintained the existing error trapping and logging mechanisms

Some Features are pending testing.

### CODE REVIEW

File: emp/custom-registry/emp_image_helper.sh [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    # Temporary file for output
+    local tmpfile
+    tmpfile=$(mktemp)
+    debug "Logs are in $tmpfile"
+    # Async pull with spinner
+    {
+        for img in $IMAGES; do
+            echo "$img" > $IMG_FILE
+            info "Getting $img"
+            $CNT_CMD pull "$img" 2>&1
```
Comment (Shubhamag12) [inline emp/custom-registry/emp_image_helper.sh:157]: we should add `--platform linux/amd64` in this command. This is due to the fact that while pulling docker images, the docker pulls the images according to the underlying architecture and as a result the pods will crash if they are of different arch., and not of linux/amd64.

**Review by Shubhamag12 — CHANGES_REQUESTED** [review]
Let us add the functionality to pull and push charts as well.
Also we should ask customers if they want us to create repositories, since we have a strict naming and if they a miss a letter or two, the EMP might fail.

**Comment by Sven27** [comment]
To test these are the formats  
(for ecr)
./emp_image_helper.sh push public.ecr.aws/o5r0j4u8 --type ecr --profile emp-dev-admin --region us-east-1
(for jfrog)
./cred.sh push artifactory.platform9.horse --type jfrog --creds ZW1wLWF1dG9tYXRpb246S2Y5UVJiKkRwTUVG --port 443
The cred need to be base64 encoded

File: emp/custom-registry/emp_image_helper.sh [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+                        echo "Registry Type: $READ_REPO"  # Debugging
+                        if [[ "$READ_REPO" != "ecr" && "$READ_REPO" != "jfrog" ]]; then
+                            echo "Please specify a correct registry type (ecr/jfrog)"
+                            exit 1
+                        fi
+                        shift 2
+                        ;;
+                    --profile)
+                        profile="$2"
+                        echo "AWS Profile: $profile"  # Debugging
```
Comment (sharma-tapas) [inline emp/custom-registry/emp_image_helper.sh:482]: use debug instead of echo for debug logs

**Review by sharma-tapas — COMMENTED** [review]
Instead of calling out JFrog just say registry.

File: emp/custom-registry/emp_image_helper.sh [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+        for img in $IMAGES; do
+            local tg
+            name_split "$img"
+            VERSION="${img##*:}"
+            echo "$VERSION"
+            if [[ "$rflag" == 1 ]]; then
+		        aws_create_repo "$3" "$4" "$1" "$2"
+	        else
+		        debug "here in jfrog"
+		        debug "tagging:$tg"
```
Comment (maheshbaviskar) [inline emp/custom-registry/emp_image_helper.sh:412]: Remove this line as this this gives error $tg before assign or move tag code above this condition.

**Comment by Sven27** [comment]
test images
Screenshot 2024-12-18 at 2 08 24 PM

**Comment by Sven27** [comment]
Screenshot 2024-12-23 at 12 33 58 PM

**Review by AnirudhPokala — APPROVED** [review]

**Review by Sven27 — APPROVED** [review]

**Pull request merged**